### ISSUE TITLE
Rule idea: extends-before-declarations / extends-before-mixins / mixins-before-declarations

### ISSUE THREAD

**Issue by niksy** [kristerkari/stylelint-scss#43]
* https://github.com/sasstools/sass-lint/blob/master/docs/rules/extends-before-declarations.md
* https://github.com/sasstools/sass-lint/blob/master/docs/rules/extends-before-mixins.md
* https://github.com/sasstools/sass-lint/blob/master/docs/rules/mixins-before-declarations.md

I think everyone who uses Sass extensively could appreciate these rules. What do you think?

**Comment by dryoma** [comment]
Thanks for the suggestion! I agree, there definitely has to be a way to force-sort stuff inside a rule, and now is a good opportunity to start discussing this matter :)

I, for one, was thinking of making it just one rule. Something similar to stylelint's [`declaration-block-properties-order`](http://stylelint.io/user-guide/rules/declaration-block-properties-order/), with array of values like `[ "mixin", "extend", "declarations", "media", "nesting" ]`.

**Comment by kristerkari** [comment]
> I, for one, was thinking of making it just one rule. Something similar to stylelint's declaration-block-properties-order, with array of values like [ "mixin", "extend", "declarations", "media", "nesting" ].

This sounds good 👍

**Comment by jeddy3** [comment]
> I, for one, was thinking of making it just one rule.

Perhaps you and the `postcss-sorting` guys can [combine forces on this one](
https://github.com/hudochenkov/postcss-sorting/issues/18#issuecomment-223791271) :) Such a rule would be a great help to the stylelint community.

Excuse the brevity... I'm on a phone.

**Comment by niksy** [comment]
@dryoma yup, when I posted this, I quickly realized this should probably work the way you suggested :D

@jeddy3 yeah, making feature set from `postcss-sorting` as `stylelint` plugin would be great, so if and when that becomes plugin, I suppose this proposed rule idea would basically be predefined [at-rule definition](https://github.com/hudochenkov/postcss-sorting#at-rules) from that new plugin?

**Comment by jeddy3** [comment]
> I suppose this proposed rule idea would basically be predefined at-rule definition from that new plugin?

Yes, I think these rule could be achieved using those at-rule options. I'm not sure they'll be a "predefined definition" though as stylelint rules tend to come without defaults i.e. each user decided whether they want mixins before extends, or vice versa.

**Comment by hudochenkov** [comment]
We are going to make a plugin for rules content order based on postcss-sorting. https://github.com/stylelint/stylelint/issues/1500#issuecomment-229994733 It will allow to define order of `@extend`, `@include`, declarations, etc.

**Comment by dryoma** [comment]
@hudochenkov would you consider making the plugin part of this plugin set? I'd personally really love to have everything related to SCSS syntax in one place.

**Comment by hudochenkov** [comment]
I didn't use SCSS, but I use SCSS-like syntax with PostCSS plugins. There is no need to limit stylelint-block-non-properties-order to this plugin set, because it can be used not only with SCSS.

**Comment by dryoma** [comment]
Well, so do rules from stylelint-scss :) But it was decided to keep the rule plugins under the same roof so that the users wouldn't have to import them one by one. That's why we've been giving the *things* the most universal names possible (`at-function` and `dollar-variable` instead of `scss-function` and `scss-variable`), and even asked the stylelint guys to make [an interface](https://github.com/stylelint/stylelint/issues/1174) for multiple-rule plugin sets.

**Comment by hudochenkov** [comment]
I made a plugin for checking these things https://github.com/hudochenkov/stylelint-declaration-block-order

For example config to specify this order: variables, extends, includes, declarations, includes with content (e. g., mixins for media queries), nested at-rules (e. g. `@media`), nested rules:

```json
{
    "plugins": [
        "stylelint-declaration-block-order"
    ],
    "rules": {
        "plugin/declaration-block-order": [
            "dollar-variables",
            {
                "type": "at-rule",
                "name": "extend",
            },
            {
                "type": "at-rule",
                "name": "include",
            },
            "declarations",
            {
                "type": "at-rule",
                "name": "include",
                "hasBlock": true
            },
            "at-rules",
            "rules"
        ]
    }
}
```

**Comment by jeddy3** [comment]
Thanks to @hudochenkov, this can now be done with [stylelint-order](https://github.com/hudochenkov/stylelint-order). stylelint-order is also the home of the rules that supersede the `declaration-block-properties-order` rule from stylelint, which will be deprecated in `stylelint@8.0.0`.

I think this issue can be closed now.

**Comment by dryoma** [comment]
TODO: add a note to the docs.
